File: mariadb-api/src/Controllers/StrategyController.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Controllers;

use Enkl\Api\Db\Database;
use Enkl\Api\Services\StrategyFulfilmentService;
use Enkl\Api\Services\StrategyMetricService;
use Enkl\Api\Services\StrategyPillarService;
use Enkl\Api\Services\StrategyService;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

/**
 * Ported from php-api's Controllers/StrategyController.php (itself ported from
 * Controllers/StrategyController.cs). Org-Admin-only management of Strategies/Pillars/Enablers/
 * Metrics + the fulfilment-matrix upsert and OrgAdmin-side read — gated by OrgAdminMiddleware ONLY
 * (see routes.php). The read-only ProjectMember surface lives in ProjectStrategyController instead.
 */
final class StrategyController extends BaseController
{
    private function strategies(): StrategyService
    {
        return new StrategyService(Database::connection());
    }

    private function pillars(): StrategyPillarService
    {
        return new StrategyPillarService(Database::connection());
    }

    private function metrics(): StrategyMetricService
    {
        return new StrategyMetricService(Database::connection());
    }

    private function fulfilment(): StrategyFulfilmentService
    {
        return new StrategyFulfilmentService(Database::connection());
    }

    public function list(Request $request, Response $response): Response
    {
        return $this->json($response, $this->strategies()->list($this->callerOrgId($request)));
    }

    public function getActive(Request $request, Response $response): Response
    {
        $active = $this->strategies()->getActive($this->callerOrgId($request));
        return $active !== null ? $this->json($response, $active) : $this->notFound($response);
    }

    public function create(Request $request, Response $response): Response
    {
        return $this->json($response, $this->strategies()->create($this->callerOrgId($request), $this->body($request)));
    }

    public function update(Request $request, Response $response, array $args): Response
    {
        $result = $this->strategies()->update($this->callerOrgId($request), $args['strategyId'], $this->body($request));
        return $result !== null ? $this->json($response, $result) : $this->notFound($response);
    }

    public function activate(Request $request, Response $response, array $args): Response
    {
        $result = $this->strategies()->activate($this->callerOrgId($request), $args['strategyId']);
        return $result !== null ? $this->json($response, $result) : $this->notFound($response);
    }

    public function delete(Request $request, Response $response, array $args): Response
    {
        $deleted = $this->strategies()->delete($this->callerOrgId($request), $args['strategyId']);
        return $deleted ? $this->noContent($response) : $this->notFound($response);
    }

    // Strategy ownership check happens implicitly: an empty tree for a foreign-org strategyId looks
    // identical to a real, empty own-org strategy — no enumeration oracle either way.
    public function getTree(Request $request, Response $response, array $args): Response
    {
        $strategies = $this->strategies()->list($this->callerOrgId($request));
        $owned = array_filter($strategies, static fn(array $s) => $s['id'] === $args['strategyId']);
        if (count($owned) === 0) {
            return $this->notFound($response);
        }
        return $this->json($response, $this->pillars()->getPillarTree($args['strategyId']));
    }

    public function createPillar(Request $request, Response $response, array $args): Response
    {
        $result = $this->pillars()->createPillar($this->callerOrgId($request), $args['strategyId'], $this->body($request));
        return $result !== null ? $this->json($response, $result) : $this->notFound($response);
    }

    public function updatePillar(Request $request, Response $response, array $args): Response
    {
        $result = $this->pillars()->updatePillar($this->callerOrgId($request), $args['pillarId'], $this->body($request));
        return $result !== null ? $this->json($response, $result) : $this->notFound($response);
    }

    public function deletePillar(Request $request, Response $response, array $args): Response
    {
        $deleted = $this->pillars()->deletePillar($this->callerOrgId($request), $args['pillarId']);
        return $deleted ? $this->noContent($response) : $this->notFound($response);
    }

    public function createEnabler(Request $request, Response $response, array $args): Response
    {
        $result = $this->pillars()->createEnabler($this->callerOrgId($request), $args['pillarId'], $this->body($request));
        return $result !== null ? $this->json($response, $result) : $this->notFound($response);
    }

    public function updateEnabler(Request $request, Response $response, array $args): Response
    {
        $result = $this->pillars()->updateEnabler($this->callerOrgId($request), $args['enablerId'], $this->body($request));
        return $result !== null ? $this->json($response, $result) : $this->notFound($response);
    }

    public function deleteEnabler(Request $request, Response $response, array $args): Response
    {
        $deleted = $this->pillars()->deleteEnabler($this->callerOrgId($request), $args['enablerId']);
        return $deleted ? $this->noContent($response) : $this->notFound($response);
    }

    public function createMetricOnPillar(Request $request, Response $response, array $args): Response
    {
        $result = $this->metrics()->create($this->callerOrgId($request), $args['pillarId'], null, $this->body($request));
        return $result !== null ? $this->json($response, $result) : $this->json($response, ['message' => 'Invalid metric request.'], 400);
    }

    public function createMetricOnEnabler(Request $request, Response $response, array $args): Response
    {
        $result = $this->metrics()->create($this->callerOrgId($request), null, $args['enablerId'], $this->body($request));
        return $result !== null ? $this->json($response, $result) : $this->json($response, ['message' => 'Invalid metric request.'], 400);
    }

    public function updateMetric(Request $request, Response $response, array $args): Response
    {
        $result = $this->metrics()->update($this->callerOrgId($request), $args['metricId'], $this->body($request));
        return $result !== null ? $this->json($response, $result) : $this->notFound($response);
    }

    public function deleteMetric(Request $request, Response $response, array $args): Response
    {
        $deleted = $this->metrics()->delete($this->callerOrgId($request), $args['metricId']);
        return $deleted ? $this->noContent($response) : $this->notFound($response);
    }

    public function recordMetricEntry(Request $request, Response $response, array $args): Response
    {
        $result = $this->metrics()->recordEntry($this->callerOrgId($request), $args['metricId'], $this->body($request));
        return $result !== null ? $this->json($response, $result) : $this->notFound($response);
    }

    public function getMetricHistory(Request $request, Response $response, array $args): Response
    {
        $result = $this->metrics()->getHistory($this->callerOrgId($request), $args['metricId']);
        return $result !== null ? $this->json($response, $result) : $this->notFound($response);
    }

    // GET, not POST — pure read, same MustChangePassword-gate-avoidance reasoning as
    // PortfolioController::getAggregate.
    public function getFulfilmentMatrix(Request $request, Response $response): Response
    {
        $query = $request->getQueryParams();
        $projectIds = array_values(array_filter(array_map('trim', explode(',', (string) ($query['projectIds'] ?? '')))));
        return $this->json($response, $this->fulfilment()->buildMatrix($this->callerOrgId($request), $projectIds));
    }

    // Logically nested under Portfolio Planner's own route namespace (the only place this is ever
    // written from) — see routes.php for the absolute path this is registered under.
    public function upsertFulfilment(Request $request, Response $response, array $args): Response
    {
        $result = $this->fulfilment()->upsert($this->callerOrgId($request), $args['projectId'], $args['pillarId'], $this->body($request));
        return $result !== null ? $this->json($response, $result) : $this->notFound($response);
    }
}

File: api/Enkl.Api/Controllers/StrategyController.cs
using Enkl.Api.Auth;
using Enkl.Api.Dtos;
using Enkl.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Enkl.Api.Controllers;

/// <summary>
/// Org-Admin-only management of Strategies/Pillars/Enablers/Metrics + the fulfilment-matrix upsert
/// and OrgAdmin-side read. OrgAdmin policy ONLY (same reasoning as PortfolioController: an admin
/// managing strategy may not personally belong to every project it's mapped against). The
/// read-only ProjectMember surface lives in ProjectStrategyController instead.
/// </summary>
[ApiController]
[Authorize(Policy = "OrgAdmin")]
[Route("api/organisations/me/strategy")]
public class StrategyController : ControllerBase
{
    private readonly StrategyService _strategies;
    private readonly StrategyPillarService _pillars;
    private readonly StrategyMetricService _metrics;
    private readonly StrategyFulfilmentService _fulfilment;

    public StrategyController(StrategyService strategies, StrategyPillarService pillars, StrategyMetricService metrics, StrategyFulfilmentService fulfilment)
    {
        _strategies = strategies;
        _pillars = pillars;
        _metrics = metrics;
        _fulfilment = fulfilment;
    }

    [HttpGet]
    public async Task<IActionResult> List()
    {
        return Ok(await _strategies.ListAsync(User.OrgId()));
    }

    [HttpGet("active")]
    public async Task<IActionResult> GetActive()
    {
        var active = await _strategies.GetActiveAsync(User.OrgId());
        return active is null ? NotFound() : Ok(active);
    }

    [HttpPost]
    public async Task<IActionResult> Create(CreateStrategyRequest request)
    {
        return Ok(await _strategies.CreateAsync(User.OrgId(), request));
    }

    [HttpPut("{strategyId:guid}")]
    public async Task<IActionResult> Update(Guid strategyId, UpdateStrategyRequest request)
    {
        var result = await _strategies.UpdateAsync(User.OrgId(), strategyId, request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpPut("{strategyId:guid}/activate")]
    public async Task<IActionResult> Activate(Guid strategyId)
    {
        var result = await _strategies.ActivateAsync(User.OrgId(), strategyId);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpDelete("{strategyId:guid}")]
    public async Task<IActionResult> Delete(Guid strategyId)
    {
        return await _strategies.DeleteAsync(User.OrgId(), strategyId) ? NoContent() : NotFound();
    }

    [HttpGet("{strategyId:guid}/tree")]
    public async Task<IActionResult> GetTree(Guid strategyId)
    {
        // Strategy ownership check happens implicitly: an empty tree for a foreign-org strategyId
        // looks identical to a real, empty own-org strategy — no enumeration oracle either way.
        var active = await _strategies.ListAsync(User.OrgId());
        if (!active.Any(s => s.Id == strategyId)) return NotFound();
        return Ok(await _pillars.GetPillarTreeAsync(strategyId));
    }

    [HttpPost("{strategyId:guid}/pillars")]
    public async Task<IActionResult> CreatePillar(Guid strategyId, CreateStrategyPillarRequest request)
    {
        var result = await _pillars.CreatePillarAsync(User.OrgId(), strategyId, request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpPut("pillars/{pillarId:guid}")]
    public async Task<IActionResult> UpdatePillar(Guid pillarId, UpdateStrategyPillarRequest request)
    {
        var result = await _pillars.UpdatePillarAsync(User.OrgId(), pillarId, request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpDelete("pillars/{pillarId:guid}")]
    public async Task<IActionResult> DeletePillar(Guid pillarId)
    {
        return await _pillars.DeletePillarAsync(User.OrgId(), pillarId) ? NoContent() : NotFound();
    }

    [HttpPost("pillars/{pillarId:guid}/enablers")]
    public async Task<IActionResult> CreateEnabler(Guid pillarId, CreateStrategyEnablerRequest request)
    {
        var result = await _pillars.CreateEnablerAsync(User.OrgId(), pillarId, request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpPut("enablers/{enablerId:guid}")]
    public async Task<IActionResult> UpdateEnabler(Guid enablerId, UpdateStrategyEnablerRequest request)
    {
        var result = await _pillars.UpdateEnablerAsync(User.OrgId(), enablerId, request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpDelete("enablers/{enablerId:guid}")]
    public async Task<IActionResult> DeleteEnabler(Guid enablerId)
    {
        return await _pillars.DeleteEnablerAsync(User.OrgId(), enablerId) ? NoContent() : NotFound();
    }

    [HttpPost("pillars/{pillarId:guid}/metrics")]
    public async Task<IActionResult> CreateMetricOnPillar(Guid pillarId, CreateStrategyMetricRequest request)
    {
        var result = await _metrics.CreateAsync(User.OrgId(), pillarId, null, request);
        return result is null ? BadRequest() : Ok(result);
    }

    [HttpPost("enablers/{enablerId:guid}/metrics")]
    public async Task<IActionResult> CreateMetricOnEnabler(Guid enablerId, CreateStrategyMetricRequest request)
    {
        var result = await _metrics.CreateAsync(User.OrgId(), null, enablerId, request);
        return result is null ? BadRequest() : Ok(result);
    }

    [HttpPut("metrics/{metricId:guid}")]
    public async Task<IActionResult> UpdateMetric(Guid metricId, UpdateStrategyMetricRequest request)
    {
        var result = await _metrics.UpdateAsync(User.OrgId(), metricId, request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpDelete("metrics/{metricId:guid}")]
    public async Task<IActionResult> DeleteMetric(Guid metricId)
    {
        return await _metrics.DeleteAsync(User.OrgId(), metricId) ? NoContent() : NotFound();
    }

    [HttpPost("metrics/{metricId:guid}/entries")]
    public async Task<IActionResult> RecordMetricEntry(Guid metricId, CreateStrategyMetricEntryRequest request)
    {
        var result = await _metrics.RecordEntryAsync(User.OrgId(), metricId, request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpGet("metrics/{metricId:guid}/entries")]
    public async Task<IActionResult> GetMetricHistory(Guid metricId)
    {
        var result = await _metrics.GetHistoryAsync(User.OrgId(), metricId);
        return result is null ? NotFound() : Ok(result);
    }

    // GET, not POST — pure read, same MustChangePassword-gate-avoidance reasoning as
    // PortfolioController.GetAggregate. projectIds is a single comma-joined string, matching
    // PortfolioController's own convention for cross-framework array-param portability.
    [HttpGet("fulfilment-matrix")]
    public async Task<IActionResult> GetFulfilmentMatrix([FromQuery] string? projectIds)
    {
        var ids = (projectIds ?? "")
            .Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
            .Select(s => Guid.TryParse(s, out var g) ? g : (Guid?)null)
            .Where(g => g.HasValue)
            .Select(g => g!.Value)
            .ToList();
        return Ok(await _fulfilment.BuildMatrixAsync(User.OrgId(), ids));
    }

    // Absolute route override — logically nested under Portfolio Planner's own namespace (the only
    // place this is ever written from, per the approved plan) rather than under this controller's
    // own /strategy prefix. A genuine mutation, so PUT (trips the MustChangePassword gate, as intended).
    [HttpPut("~/api/organisations/me/portfolio/projects/{projectId:guid}/strategy-fulfilment/{pillarId:guid}")]
    public async Task<IActionResult> UpsertFulfilment(Guid projectId, Guid pillarId, UpsertProjectPillarFulfilmentRequest request)
    {
        var result = await _fulfilment.UpsertAsync(User.OrgId(), projectId, pillarId, request);
        return result is null ? NotFound() : Ok(result);
    }
}
